North-Seattle-College/ad440-winter2021-thursday-repo
# [BUG] users/userId/tasks/taskId API logging and error handling Issue #83

### Resolves #83 

This code completes the implementation of the Azure function for the `/users/user_id/tasks/task_id endpoint`
The function will be deployed to the cloud following the approval and merge into the development branch
It doesn't require configuration to be run on the cloud
### To test locally:
Make sure you have the Azure functions extension for VSCode
Pull the latest iteration of this branch
Install the ODBC driver necessary to connect to the database (pyodbc or pypyodbc)
Configure your venv and vscode files to set up the appropriate runtime for this Azure function
In local.settings.json at the root of the functions folder, enter in the values for the keys below in the following format: { "Values": { "KEY": "VALUE" } }
ENV_DATABASE_NAME (name of the database)
ENV_DATABASE_PASSWORD
ENV_DATABASE_SERVER (this is the server location)
ENV_DATABASE_USERNAME

In VSCode, click `F5 `or `run->Start Debugging `in the functions folder to fire up the function
After the function has started, input the url provided in terminal window into Postman, and test the functionality by selecting the appropriate method and url parameters and/or json body, like so:

**GET method**
Pass the userId and taskId as url parameters, JSON body is not required
<img width="1009" alt="Screen Shot 2021-01-31 at 5 20 07 PM" src="https://user-images.githubusercontent.com/46657216/106405856-5be38080-63ec-11eb-98be-8da841847a09.png">


**PUT method**:
Pass the userId and taskId as url parameters, JSON body IS required to include `title`, `description`, `dueDate`, `completed`, `completedDate` (all five updatable fields need to be present, all other JSON fields will be ignored), you may pass more by cutting and pasting from GET results
You may use this sample JSON for update testing:
```
{
    "userId": 16,
    "taskId": 24,
    "title": "My Question",
    "description": "Decide on if to be or not to be",
    "dueDate": null,
    "completed": 1,
    "completedDate":"21/01/07 05:46:38"
}
```
Note that the in-built `datetime `conversion will only work for legitimate `yy/mm/dd hh:mm:ss` If you attempt to update the due date to occur on 20/13/32, you will receive an error.
**Be sure to add the data in the `"yy/mm/dd hh:mm:ss"` string format**, otherwise you will receive an error.

successful request results:
<img width="998" alt="Screen Shot 2021-01-31 at 5 20 47 PM" src="https://user-images.githubusercontent.com/46657216/106406035-e62be480-63ec-11eb-9f94-33da60aa76da.png">
**PATCH method**
Same instructions as for PUT, userId and taskId are required, JSON body may contain one to five key-value pairs excluding userId and taskId. If "Completed" is included, it may not be null per db specifications.
<img width="991" alt="Screen Shot 2021-01-31 at 5 35 52 PM" src="https://user-images.githubusercontent.com/46657216/106406150-4458c780-63ed-11eb-9fa5-52f75828842b.png">

PUT and PATCH results: 
<img width="992" alt="Screen Shot 2021-01-31 at 5 36 19 PM" src="https://user-images.githubusercontent.com/46657216/106406186-5fc3d280-63ed-11eb-9bc7-ec6b43a785de.png">


**DELETE method**
Pass the userId and taskId as url parameters, JSON body is not required.
_Please DO NOT delete the record with userId = 16, taskId = 24
taskId = 25-26 for userId =16 are available for the delete functionality testing.
If the requested taskId is not in the db, you will receive an error message:
<img width="1000" alt="Screen Shot 2021-01-28 at 2 44 17 PM" src="https://user-images.githubusercontent.com/46657216/106208145-c6d85180-6177-11eb-85fe-a51e2bbcb2aa.png">
